### PR TITLE
systemd, osutil: rework systemd logs in preparation for services commands

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

File: systemd/systemd.go [hunk 1]
```diff
@@ -62,21 +60,19 @@ func run(args ...string) ([]byte, error) {
...
 
-// jctl calls journalctl to get the JSON logs of the given services, wrapping the error if any.
-func jctl(svcs []string) ([]byte, error) {
-	cmd := []string{"journalctl", "-o", "json"}
-
-	for i := range svcs {
-		cmd = append(cmd, "-u", svcs[i])
+// jctl calls journalctl to get the JSON logs of the given services.
+func jctl(svcs []string, n string, follow bool) (io.ReadCloser, error) {
+	args := make([]string, 0, 2*len(svcs)+4)
```
Comment (stolowski) [inline systemd/systemd.go]: Just curious, is this because we know we will have at least that many lines? Or totally out of blue number?

Comment (chipaca) [inline systemd/systemd.go]: I like making numbers up, but in this case I know we'll have at most that many lines, at at least that many minus one¹, so doing it this way is not much more complex than starting with a nil slice, and can save a bunch of memory (as append doubles the size of the backing array every time it runs out). The added complexity of sizing it exactly right didn't seem worth it.

1. although looking at the maths now I might have not updated the number after the last tweak, I'll double-check.

Comment (stolowski) [inline systemd/systemd.go]: Ah, ok, I'm glad I've asked. It's not critical, but a comment would be nice indeed.

File: osutil/exec.go [hunk 2]
```diff
@@ -236,3 +237,47 @@ func RunAndWait(argv []string, env []string, timeout time.Duration, tomb *tomb.T
...
+	}
+	return n, err
+}
+
+// StreamCommand runs a the named program with the given arguments,
+// streaming its standard output over the returned io.ReadCloser.
+//
+// The program will run until EOF is reached (at which point the
+// ReadCloser is closed), or until the ReadCloser is explicitly closed.
+func StreamCommand(name string, args ...string) (io.ReadCloser, error) {
```
Comment (mvo5) [inline osutil/exec.go]: Nice!

File: systemd/systemd.go [hunk 3]
```diff
@@ -395,15 +367,27 @@ func (l Log) Message() string {
...
-	if sid, ok := l["SYSLOG_IDENTIFIER"].(string); ok {
+	if sid, ok := l["SYSLOG_IDENTIFIER"]; ok {
 		return sid
 	}
 
 	return "-"
 }
 
+// PID is the pid of the client pid, if any; otherwise, "-".
+func (l Log) PID() string {
```
Comment (mvo5) [inline systemd/systemd.go]: PID is spelled Pid in e.g. os.ProcessState - maybe we should follow this convention.

**Pull request merged**